File: functions/controllers/thumbnailController.js
/**
 * 缩略图控制器
 * 处理缩略图生成相关业务逻辑
 */

import { decodeImage, encodeImage, resizeImage, detectImageFormat } from '../utils/jsquash.js';
import { 
	MAX_IMAGE_SIZE, 
	DEFAULT_QUALITY, 
	CACHE_MAX_AGE, 
	SUPPORTED_FORMATS,
	THUMBNAIL_MAX_SIZE,
	VALID_FIT_METHODS,
	DEFAULT_THUMBNAIL_WIDTH,
	DEFAULT_THUMBNAIL_HEIGHT,
	DEFAULT_THUMBNAIL_FIT
} from '../constants.js';
import { getImageData } from './imageController.js';

/**
 * 验证并解析宽度和高度参数
 * 支持自适应：只提供 width 时高度按比例计算，只提供 height 时宽度按比例计算
 * 默认值：width=200, height=200
 * @param {string} width - 宽度参数
 * @param {string} height - 高度参数
 * @param {number} originalWidth - 原图宽度
 * @param {number} originalHeight - 原图高度
 * @returns {Object} 宽度和高度值
 */
function validateDimensions(width, height, originalWidth, originalHeight) {
	const hasWidth = width && width.trim() !== '';
	const hasHeight = height && height.trim() !== '';
	
	let widthValue, heightValue;
	
	if (hasWidth && hasHeight) {
		// 两者都提供，使用指定值
		widthValue = parseInt(width, 10);
		heightValue = parseInt(height, 10);
	} else if (hasWidth) {
		// 只提供宽度，高度按比例自适应
		widthValue = parseInt(width, 10);
		if (originalWidth > 0 && originalHeight > 0) {
			heightValue = Math.round((widthValue * originalHeight) / originalWidth);
		} else {
			heightValue = DEFAULT_THUMBNAIL_HEIGHT;
		}
	} else if (hasHeight) {
		// 只提供高度，宽度按比例自适应
		heightValue = parseInt(height, 10);
		if (originalWidth > 0 && originalHeight > 0) {
			widthValue = Math.round((heightValue * originalWidth) / originalHeight);
		} else {
			widthValue = DEFAULT_THUMBNAIL_WIDTH;
		}
	} else {
		// 两者都未提供，使用默认值
		widthValue = DEFAULT_THUMBNAIL_WIDTH;
		heightValue = DEFAULT_THUMBNAIL_HEIGHT;
	}
	
	// 验证数值有效性
	if (isNaN(widthValue) || isNaN(heightValue) || widthValue <= 0 || heightValue <= 0 || 
		widthValue > THUMBNAIL_MAX_SIZE || heightValue > THUMBNAIL_MAX_SIZE) {
		throw new Error(`Width and height must be between 1 and ${THUMBNAIL_MAX_SIZE}`);
	}
	
	return { widthValue, heightValue };
}

/**
 * 验证裁剪模式
 * 默认值：cover（等比例缩放，填满目标尺寸，禁止变形）
 * @param {string} fitMethod - 裁剪模式
 * @returns {string} 验证后的裁剪模式
 */
function validateFitMethod(fitMethod) {
	// 如果未提供参数，使用默认值
	if (!fitMethod) {
		return DEFAULT_THUMBNAIL_FIT;
	}
	
	const fitMethodValue = fitMethod.toLowerCase();
	if (!VALID_FIT_METHODS.includes(fitMethodValue)) {
		throw new Error(`Invalid fit method. Use one of: ${VALID_FIT_METHODS.join(', ')}`);
	}
	return fitMethodValue;
}

/**
 * 验证输出格式
 * @param {string} format - 输出格式
 * @returns {string} 验证后的格式
 */
function validateFormat(format) {
	const formatLower = format.toLowerCase();
	if (!SUPPORTED_FORMATS.has(formatLower)) {
		throw new Error(`Unsupported format: ${format}. Supported: ${Array.from(SUPPORTED_FORMATS).join(', ')}`);
	}
	return formatLower;
}

/**
 * 验证质量参数
 * 默认值：85
 * @param {string} quality - 质量参数
 * @returns {number} 质量值
 */
function validateQuality(quality) {
	// 如果未提供参数，使用默认值
	if (!quality) return DEFAULT_QUALITY;
	
	const qualityValue = parseInt(quality, 10);
	if (isNaN(qualityValue) || qualityValue < 1 || qualityValue > 100) {
		throw new Error('Quality must be between 1 and 100');
	}
	
	return qualityValue;
}

/**
 * 处理缩略图生成
 * 注意：缩略图统一使用 webp 格式输出
 * @param {Object} c - Hono 上下文对象
 * @returns {Promise<Response>} 缩略图响应
 */
export async function generateThumbnail(c) {
	try {
		// 解析查询参数
		const width = c.req.query('width') || c.req.query('w');
		const height = c.req.query('height') || c.req.query('h');
		const fitMethod = c.req.query('fit');
		const quality = c.req.query('quality') || c.req.query('q');
		// 缩略图统一使用 webp 格式，忽略用户传入的 format 参数
		const formatValue = 'webp';
		
		// 验证基础参数
		let fitMethodValue, qualityValue;
		try {
			fitMethodValue = validateFitMethod(fitMethod);
			qualityValue = validateQuality(quality);
		} catch (e) {
			return c.json({ error: e.message }, 400);
		}
		
		// 获取图片数据
		let imageData;
		try {
			imageData = await getImageData(c);
		} catch (e) {
			return c.json({ error: e.message }, 400);
		}
		
		// 限制图片大小
		if (imageData.byteLength > MAX_IMAGE_SIZE) {
			return c.json({ error: `Image too large. Maximum size: ${MAX_IMAGE_SIZE / 1024 / 1024}MB` }, 400);
		}
		
		// 检测原始图片格式
		const originalFormat = detectImageFormat(imageData);
		if (originalFormat === 'unknown') {
			return c.json({ error: 'Unsupported image format' }, 400);
		}
		
		// 解码图片
		let decodedImageData;
		try {
			decodedImageData = await decodeImage(imageData, originalFormat);
		} catch (e) {
			return c.json({ error: `Failed to decode image: ${e.message}` }, 400);
		}
		
		// 获取原图尺寸，用于计算自适应尺寸
		const originalWidth = decodedImageData.width;
		const originalHeight = decodedImageData.height;
		
		// 验证并计算目标尺寸（支持自适应）
		let widthValue, heightValue;
		try {
			({ widthValue, heightValue } = validateDimensions(width, height, originalWidth, originalHeight));
		} catch (e) {
			return c.json({ error: e.message }, 400);
		}
		
		// 生成缩略图（调整尺寸）
		let thumbnailData;
		try {
			thumbnailData = await resizeImage(decodedImageData, {
				width: widthValue,
				height: heightValue,
				fitMethod: fitMethodValue
			});
		} catch (e) {
			return c.json({ error: `Failed to resize image: ${e.message}` }, 500);
		}
		
		// 编码为目标格式
		let thumbnailImage;
		try {
			thumbnailImage = await encodeImage(thumbnailData, formatValue, qualityValue);
		} catch (e) {
			return c.json({ error: `Failed to encode image: ${e.message}` }, 500);
		}
		
		// 返回缩略图（统一使用 webp 格式）
		return c.body(thumbnailImage, 200, {
			'Content-Type': 'image/webp',
			'Cache-Control': `public, max-age=${CACHE_MAX_AGE}, immutable`,
			'X-Thumbnail-Size': `${widthValue}x${heightValue}`,
			'X-Original-Size': imageData.byteLength.toString(),
			'X-Thumbnail-Size-Bytes': thumbnailImage.byteLength.toString(),
			'X-Format': 'webp' // 标识输出格式
		});
		
	} catch (e) {
		return c.json({ error: `Failed to generate thumbnail: ${e.message}` }, 500);
	}
}

File: server/controllers/thumbnail_controller.py
"""
缩略图控制器
处理缩略图生成相关业务逻辑
"""

from flask import request, jsonify, send_file
from PIL import Image
import pillow_avif
from io import BytesIO
from .image_controller import get_image_data, MAX_IMAGE_SIZE, DEFAULT_QUALITY, CACHE_MAX_AGE

# 缩略图最大尺寸
THUMBNAIL_MAX_SIZE = 5000

# 缩略图默认宽度
DEFAULT_THUMBNAIL_WIDTH = 200

# 缩略图默认高度
DEFAULT_THUMBNAIL_HEIGHT = 200

# 缩略图默认裁剪模式
DEFAULT_THUMBNAIL_FIT = 'cover'

# 有效的裁剪模式
VALID_FIT_METHODS = ['cover', 'contain']


def validate_dimensions(width, height, original_width, original_height):
	"""
	验证并解析宽度和高度参数
	支持自适应：只提供 width 时高度按比例计算，只提供 height 时宽度按比例计算
	:param width: 宽度参数
	:param height: 高度参数
	:param original_width: 原图宽度
	:param original_height: 原图高度
	:return: 宽度和高度值
	"""
	has_width = width and width.strip() != ''
	has_height = height and height.strip() != ''
	
	if has_width and has_height:
		# 两者都提供，使用指定值
		width_value = int(width)
		height_value = int(height)
	elif has_width:
		# 只提供宽度，高度按比例自适应
		width_value = int(width)
		if original_width > 0 and original_height > 0:
			height_value = round((width_value * original_height) / original_width)
		else:
			height_value = DEFAULT_THUMBNAIL_HEIGHT
	elif has_height:
		# 只提供高度，宽度按比例自适应
		height_value = int(height)
		if original_width > 0 and original_height > 0:
			width_value = round((height_value * original_width) / original_height)
		else:
			width_value = DEFAULT_THUMBNAIL_WIDTH
	else:
		# 两者都未提供，使用默认值
		width_value = DEFAULT_THUMBNAIL_WIDTH
		height_value = DEFAULT_THUMBNAIL_HEIGHT
	
	# 验证数值有效性
	if width_value <= 0 or height_value <= 0 or \
		width_value > THUMBNAIL_MAX_SIZE or height_value > THUMBNAIL_MAX_SIZE:
		raise ValueError(f'Width and height must be between 1 and {THUMBNAIL_MAX_SIZE}')
	
	return width_value, height_value


def validate_fit_method(fit_method):
	"""
	验证裁剪模式
	:param fit_method: 裁剪模式
	:return: 验证后的裁剪模式
	"""
	if not fit_method:
		return DEFAULT_THUMBNAIL_FIT
	
	fit_method_value = fit_method.lower()
	if fit_method_value not in VALID_FIT_METHODS:
		raise ValueError(f'Invalid fit method. Use one of: {", ".join(VALID_FIT_METHODS)}')
	
	return fit_method_value


def validate_quality(quality):
	"""
	验证质量参数
	:param quality: 质量参数
	:return: 质量值
	"""
	if not quality:
		return DEFAULT_QUALITY
	
	try:
		quality_value = int(quality)
		if quality_value < 1 or quality_value > 100:
			raise ValueError('Quality must be between 1 and 100')
		return quality_value
	except (ValueError, TypeError):
		raise ValueError('Quality must be between 1 and 100')


def generate_thumbnail():
	"""
	处理缩略图生成
	注意：缩略图统一使用 webp 格式输出
	:return: Flask Response 对象
	"""
	try:
		# 解析查询参数
		width = request.args.get('width') or request.args.get('w')
		height = request.args.get('height') or request.args.get('h')
		fit_method = request.args.get('fit')
		quality = request.args.get('quality') or request.args.get('q')
		
		# 缩略图统一使用 webp 格式
		format_value = 'webp'
		
		# 验证基础参数
		try:
			fit_method_value = validate_fit_method(fit_method)
			quality_value = validate_quality(quality)
		except ValueError as e:
			return jsonify({'error': str(e)}), 400
		
		# 获取图片数据
		try:
			image_data = get_image_data()
		except ValueError as e:
			return jsonify({'error': str(e)}), 400
		
		# 限制图片大小
		if len(image_data) > MAX_IMAGE_SIZE:
			return jsonify({
				'error': f'Image too large. Maximum size: {MAX_IMAGE_SIZE / 1024 / 1024}MB'
			}), 400
		
		# 打开图片
		try:
			image = Image.open(BytesIO(image_data))
			original_width, original_height = image.size
		except Exception as e:
			return jsonify({'error': f'Failed to open image: {str(e)}'}), 400
		
		# 验证并计算目标尺寸（支持自适应）
		try:
			width_value, height_value = validate_dimensions(
				width, height, original_width, original_height
			)
		except ValueError as e:
			return jsonify({'error': str(e)}), 400
		
		# 生成缩略图（调整尺寸）
		try:
			if fit_method_value == 'contain':
				# 保持宽高比，完整显示
				ratio = min(width_value / original_width, height_value / original_height)
				new_width = int(original_width * ratio)
				new_height = int(original_height * ratio)
				image = image.resize((new_width, new_height), Image.Resampling.LANCZOS)
			elif fit_method_value == 'cover':
				# 保持宽高比，填满目标尺寸
				ratio = max(width_value / original_width, height_value / original_height)
				new_width = int(original_width * ratio)
				new_height = int(original_height * ratio)
				image = image.resize((new_width, new_height), Image.Resampling.LANCZOS)
				# 居中裁剪
				left = (new_width - width_value) // 2
				top = (new_height - height_value) // 2
				image = image.crop((left, top, left + width_value, top + height_value))
		except Exception as e:
			return jsonify({'error': f'Failed to resize image: {str(e)}'}), 500
		
		# 编码为目标格式（webp）
		output = BytesIO()
		image.save(output, format='WEBP', quality=quality_value)
		output.seek(0)
		
		# 返回缩略图
		response = send_file(
			output,
			mimetype='image/webp',
			headers={
				'Cache-Control': f'public, max-age={CACHE_MAX_AGE}, immutable',
				'X-Thumbnail-Size': f'{width_value}x{height_value}',
				'X-Original-Size': str(len(image_data)),
				'X-Thumbnail-Size-Bytes': str(output.tell()),
				'X-Format': 'webp'
			}
		)
		return response
		
	except Exception as e:
		return jsonify({'error': f'Failed to generate thumbnail: {str(e)}'}), 500

